### ISSUE TITLE
Add dynamic flow reinitialization

### ISSUE THREAD

**Issue by matkob** [akka/akka#31295]
Hi
I have a use case where I need to reinitialize a flow instance with new parameters in already running graph. Specifically I have a WebSocket client that might get redirected to a different endpoint based on incoming messages. This kind of behaviour looks similarly to how `RestartFlow` works but as far as I know there is no way to dynamically change what next flow looks like (after restart).

My proposal is to have a "dynamic flow" which would generalize `RestartFlow`'s behavior - it would accept a `trigger` of a type `Input => Option[Flow]` (simplified) that would be executed on every element. If `trigger` returns a Flow, it takes the old one's place.

I've already implemented a PoC using Akka's internal API and am curious if you'd be interested in such new component. If so, I'd be more than happy to contribute. Also bare in mind `trigger` signature is just a proposal which works for me but could/should be further discussed.

Let me know what you think about it

**Comment by johanandren** [comment]
Sounds like something you could compose with existing stages?

Something like:
```scala
Flow.map { elem => if (something) throw A(elem) else elem }
  .recoverWith { case A(elem) if (shouldRetry(elem)) => someNewFlow }
```

**Comment by johanandren** [comment]
Some clever use of a few other more advanced operators could do the trick though: 

```scala
Flow[Int].splitWhen(_ % 10 == 0)
  .prefixAndTail(1)
  .flatMapConcat { case (splitElem, rest) =>
    val newFlow = Flow[Int] // based on splitElem
    rest.via(newFlow) // possibly including splitElem through rest.prefix(splitElem)
  }
```

**Comment by matkob** [comment]
Interesting! I wouldn't have figured it out.
I ran your code locally and it works as expected, the only thing I had to add is `concatSubstreams` at the end so it has the shape of `Flow`, not `SubFlow`. Also checked and confirmed all the subflows complete when a new one gets created so I assume that will work perfectly fine with my resource-heavy use case (wss clients).

I think this issue can be closed now, thanks a lot for your support!
